hack/generator/pkg/conversions/property_assignment_function.go

```diff
@@ -295,33 +324,60 @@ func (fn *PropertyAssignmentFunction) createConversions(receiver astmodel.TypeDe
 			typeDescription.String())
 	}
 
-	receiverProperties := fn.createPropertyMap(receiverContainer)
-	otherProperties := fn.createPropertyMap(otherContainer)
+	var sourceProperties map[string]*astmodel.PropertyDefinition
+	var sourceFunctions map[string]astmodel.Function
+	var destinationProperties map[string]*astmodel.PropertyDefinition
+
+	switch fn.direction {
+	case ConvertFrom:
+		sourceProperties = fn.createPropertyMap(otherProperties)
+		sourceFunctions = fn.createFunctionMap(otherFunctions)
+		destinationProperties = fn.createPropertyMap(receiverProperties)
+	case ConvertTo:
+		sourceProperties = fn.createPropertyMap(receiverProperties)
+		sourceFunctions = fn.createFunctionMap(receiverFunctions)
+		destinationProperties = fn.createPropertyMap(otherProperties)
+	default:
+		panic(fmt.Sprintf("unexpected conversion direction %q", fn.direction))
+	}
 
 	// Flag receiver name as used
 	fn.knownLocals.Add(receiver.Name().Name())
 
-	for _, receiverProperty := range receiverProperties {
-		otherProperty, ok := otherProperties[receiverProperty.PropertyName()]
-		//TODO: Handle renames
+	for destinationName, destinationProperty := range destinationProperties {
+		sourceProperty, ok := sourceProperties[destinationName]
+		//TODO: Handle property renames
 		if ok {
-			var conv StoragePropertyConversion
-			var err error
-			switch fn.direction {
-			case ConvertFrom:
-				conv, err = fn.createPropertyConversion(otherProperty, receiverProperty)
-			case ConvertTo:
-				conv, err = fn.createPropertyConversion(receiverProperty, otherProperty)
-			default:
-				panic(fmt.Sprintf("unexpected conversion direction %q", fn.direction))
+			// Generate a conversion from one property to another
+			conv, err := fn.createPropertyConversion(sourceProperty, destinationProperty)
+			if err != nil {
+				// An error was returned, we abort creating conversions for this object
+				return errors.Wrapf(
+					err,
+					"creating conversion setting property %q from property %q",
+					destinationProperty.PropertyName(),
+					sourceProperty.PropertyName())
+			} else if conv != nil {
+				// A conversion was created, keep it for later
+				fn.conversions[destinationName] = conv
 			}
+			continue
+		}
 
+		sourceFunction, ok := sourceFunctions[destinationName]
+		if ok {
+			// Generate a conversion to assign the property from the function
+			conv, err := fn.createFunctionConversion(sourceFunction, destinationProperty)
 			if err != nil {
 				// An error was returned, we abort creating conversions for this object
-				return errors.Wrapf(err, "creating conversion for property %q of %q", receiverProperty.PropertyName(), receiver.Name())
+				return errors.Wrapf(
+					err,
+					"creating conversion setting property %q from function %s()",
+					destinationName,
+					sourceFunction.Name())
```

Comment:
These two chunks (and the `createPropertyConversion` and `createFunctionConversion` methods below) are very similar. Would it be feasible to make wrappers for sourceProperty and sourceFunction to make them expose the same interface (where one reads from a selector and the other reads from a callexpr)  and then use that in a single `createPropertyConversion` method?

Reply:
I've created two wrapper end points - `ReadableConversionEndpoint` and `WritableConversionEndpoint` that achieve a nice reduction in duplication.